### PR TITLE
fixes #248 - make apoc.export.cypher support the format for cypher-shell

### LINKED ISSUE #248: make apoc.export.cypher support the format for cypher-shell

### PR DESCRIPTION
**Pull request by AngeloBusato**

### CODE REVIEW

File: docs/overview.adoc [hunk 1]
```diff
@@ -212,8 +212,11 @@ RETURN apoc.meta.isType(n.age,"INTEGER") as ageType
 
 === Export to Cypher Script
 
-Data is exported as cypher statements (for neo4j-shell, and partly apoc.cypher.runFile to the given file.
-
+Data is exported as cypher statements to the given file.
+It is possible to choose between two export formats:
+    * `neo4j-shell`: for neo4j shell and partly apoc.cypher.runFile
+    * `cypher-shell`: for cypher shell
+To change the export format, you have to set it on the config params like `{format : cypher-shell}`. By default the format is neo4j-shell.
```
Comment (jexp) [inline docs/overview.adoc]: "cypher-shell"

File: docs/overview.adoc [hunk 2]
```diff
@@ -212,8 +212,11 @@ RETURN apoc.meta.isType(n.age,"INTEGER") as ageType
 
 === Export to Cypher Script
 
-Data is exported as cypher statements (for neo4j-shell, and partly apoc.cypher.runFile to the given file.
-
+Data is exported as cypher statements to the given file.
+It is possible to choose between two export formats:
+    * `neo4j-shell`: for neo4j shell and partly apoc.cypher.runFile
```
Comment (jexp) [inline docs/overview.adoc]: "neo4j-shell"

**Comment by jexp** [comment]
@AngeloBusato ping :)

**Comment by AngeloBusato** [comment]
@jexp we fixed this morning, sorry we thought GH notified to you about the changes

**Comment by jexp** [comment]
But it's not fixed? Or did I miss something. 
My comments haven't been really addressed?

File: docs/overview.adoc [hunk 3]
```diff
@@ -212,8 +212,11 @@ RETURN apoc.meta.isType(n.age,"INTEGER") as ageType
 
 === Export to Cypher Script
 
-Data is exported as cypher statements (for neo4j-shell, and partly apoc.cypher.runFile to the given file.
-
+Data is exported as cypher statements to the given file.
+It is possible to choose between two export formats:
+    * "neo4j-shell": for neo4j shell and partly apoc.cypher.runFile
+    * "cypher-shell": for cypher shell
```
Comment (AngeloBusato) [inline docs/overview.adoc]: @jexp this are our changes :)

File: docs/overview.adoc [hunk 4]
```diff
@@ -212,8 +212,11 @@ RETURN apoc.meta.isType(n.age,"INTEGER") as ageType
...
+To change the export format, you have to set it on the config params like `{format : "cypher-shell"}`. By default the format is neo4j-shell.
```
Comment (AngeloBusato) [inline docs/overview.adoc]: @jexp these are our changes :)

**Comment by jexp** [comment]
@AngeloBusato and my other comments?

As I said before I want that we pass in a Format via the ExportConfig to the Exporter, which is then asked for begin, commit, await instructions to be written out.

And we should have a Neo4jShellFormat, CypherShellFormat and NullFormat.

Later we'll add other things like parameter support. etc.

You'll also have to rebase as @albertodelazzari's PR just went in.

**Comment by AngeloBusato** [comment]
@jexp in the nullFormat what do we expect?

**Comment by jexp** [comment]
The null format will not have any begin and commit. Just only Cypher statements.

**Comment by AngeloBusato** [comment]
ok perfect.

**Comment by AngeloBusato** [comment]
@jexp Done :)

File: src/main/java/apoc/export/cypher/CypherShellSubGraphExporter.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
+package apoc.export.cypher;
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/CypherShellSubGraphExporter.java]: @AngeloBusato ping

File: src/main/java/apoc/export/cypher/Neo4jShellSubGraphExporter.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * Outputs indexes and constraints at the beginning as their own transactions
+ */
+public class Neo4jShellSubGraphExporter extends CypherShellSubGraphExporter implements MultiStatementCypherSubGraphExporter {
+
+    public Neo4jShellSubGraphExporter(SubGraph graph) {
+        super(graph);
+    }
+
+    @Override
+    protected void begin(PrintWriter out) {
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/Neo4jShellSubGraphExporter.java]: @AngeloBusato ping

**Comment by AngeloBusato** [comment]
@jexp ok now we fix these changes.

**Comment by AngeloBusato** [comment]
@jexp fixed

File: docs/overview.adoc [hunk 7]
```diff
@@ -212,8 +212,12 @@ RETURN apoc.meta.isType(n.age,"INTEGER") as ageType
 
 === Export to Cypher Script
 
-Data is exported as cypher statements (for neo4j-shell, and partly apoc.cypher.runFile to the given file.
-
+Data is exported as cypher statements to the given file.
+It is possible to choose between three export formats:
+    * "neo4j-shell": for neo4j shell and partly apoc.cypher.runFile
+    * "cypher-shell": for cypher shell
+    * "null-format": doesn't output begin / commit / await just plain cypher
```
Comment (jexp) [inline docs/overview.adoc]: please call it "plain"

File: src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java [hunk 8]
```diff
@@ -18,26 +14,26 @@
...
+    Set<String> indexNames        = new LinkedHashSet<>();
     Set<String> indexedProperties = new LinkedHashSet<>();
     private long artificialUniques = 0;
+    private ExportFormat exportFormat = ExportFormat.NEO4J_SHELL;
 
-    public MultiStatementCypherSubGraphExporter(SubGraph graph) {
+    public MultiStatementCypherSubGraphExporter(SubGraph graph, ExportFormat exportFormat) {
         this.graph = graph;
         uniqueConstraints = gatherUniqueConstraints(indexNames, indexedProperties);
+        this.exportFormat = exportFormat;
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java]: null check? only set it if the passed in format != null

File: src/main/java/apoc/export/util/ExportFormat.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @since 06-04-2017
+ */
+public enum ExportFormat {
+
+    NEO4J_SHELL("neo4j-shell", String.format("commit%n"), String.format("begin%n"), String.format("schema await%n")),
+    CYPHER_SHELL("cypher-shell", String.format(":commit%n"), String.format(":begin%n"), ""),
+    NULL_FORMAT("null-format", "", "", "");
+
+    private String format;
```
Comment (jexp) [inline src/main/java/apoc/export/util/ExportFormat.java]: make these final

File: src/main/java/apoc/export/util/ExportFormat.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+    private String commit;
+
+    private String begin;
+
+    private String schemaAwait;
+
+    ExportFormat(String format) {
+        this.format = format;
```
Comment (jexp) [inline src/main/java/apoc/export/util/ExportFormat.java]: then all others are null?

File: src/main/java/apoc/export/util/ExportFormat.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    public static final ExportFormat fromString(String format) {
+        for (ExportFormat exportFormat : ExportFormat.values()) {
+            if (exportFormat.format.equalsIgnoreCase(format)) {
+                return exportFormat;
+            }
+        }
+        return NEO4J_SHELL;
+    }
+
+    public String begin(){
```
Comment (jexp) [inline src/main/java/apoc/export/util/ExportFormat.java]: probably better with overriden methods?

File: src/main/java/apoc/export/util/ExportFormat.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    ExportFormat(String format, String commit, String begin, String schemaAwait) {
+        this.format = format;
+        this.begin = begin;
+        this.commit = commit;
+        this.schemaAwait = schemaAwait;
+    }
+
+    public static final ExportFormat fromString(String format) {
+        for (ExportFormat exportFormat : ExportFormat.values()) {
```
Comment (jexp) [inline src/main/java/apoc/export/util/ExportFormat.java]: null check of "format" before

File: src/test/java/apoc/export/cypher/ExportCypherTest.java [hunk 13]
```diff
@@ -67,7 +70,50 @@
             "begin%n" +
             "DROP CONSTRAINT ON (node:`UNIQUE IMPORT LABEL`) ASSERT node.`UNIQUE IMPORT ID` IS UNIQUE;%n" +
             "commit");
-private static final Map<String, Object> exportConfig = Collections.singletonMap("separateFiles", true);
+
+    private static final String EXPECTED_CYPHER_SHELL = String.format(":begin%n" +
```
Comment (jexp) [inline src/test/java/apoc/export/cypher/ExportCypherTest.java]: don't copy & paste it, combine it from the same fragements that are then just joined with the proper begin/commit/await pieces.

File: docs/overview.adoc (hunk 7 above)
Comment (AngeloBusato) [inline docs/overview.adoc]: @jexp "plain" not "plain-format" right?

**Comment by AngeloBusato** [comment]
@jexp Done

File: src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java [hunk 14]
```diff
@@ -18,26 +14,26 @@
 import java.lang.reflect.Array;
 import java.util.*;
 
-/**
- * Idea is to lookup nodes for relationships via a unqiue index
- * either one inherent to the original node, or a artificial one that indexes the original node-id
- * and which is removed after the import.
- * <p>
- * Outputs indexes and constraints at the beginning as their own transactions
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java]: why was this comment removed?

**Comment by jexp** [comment]
Merged manually

**Pull request closed without merging**